### PR TITLE
Allow linking of static frameworks/libraries into dynamic frameworks

### PR DESCRIPTION
**Pull request by ollieatkinson**
Resolves https://github.com/tuist/tuist/issues/175

### Short description 📝

Tuist prevents dynamic frameworks from linking static libraries or frameworks. This was reported [here](https://github.com/tuist/tuist/issues/172). This pull request lifts those linter warnings and ensure's that the frameworks are allowed to link static dependencies as well as any transitive static dependencies.

### Solution 📦

When we traverse the graph for linkable dependencies we used to traverse the whole graph for static products and include them at the top level bundle. 

We now traverse the graph and stop looking for any more static products when we reach a dynamic framework, and the work is then handed off to the framework to decide what it links during it's pass of linkableDependencies.

I created a fixture called `ios_app_with_framework_linking_static_framework` which demonstrates the logic described above.

Todo:

- [x] Update changelog
- [x] Update fixture Readme
- [x] Create fixture test to ensure test compiles

### CODE REVIEW

File: Sources/TuistKit/Graph/Graph.swift [hunk 1]
```diff
@@ -311,7 +311,7 @@ extension Graph {
     }
 
     // Traverse the graph from the target node using DFS and return all results passing the test.
-    internal func findAll<T: GraphNode>(targetNode: TargetNode, test: (T) -> Bool) -> Set<T> {
+    internal func findAll<T: GraphNode>(targetNode: TargetNode, test: (T) -> Bool, doNotTraverseWhen: (T) -> Bool = { _ in false }) -> Set<T> {
```
Comment (pepibumur) [inline Sources/TuistKit/Graph/Graph.swift]: Nitpick: What about calling the argument `skip`?

Comment (ollieatkinson) [reply Sources/TuistKit/Graph/Graph.swift]: I called it skip first, and I changed it as I wasn't sure it was obvious enough - maybe I should have trusted my instincts :D

File: Tests/TuistKitTests/Graph/GraphTests.swift [hunk 2]
```diff
@@ -88,19 +88,39 @@ final class GraphTests: XCTestCase {
...
+        XCTAssertEqual(got.count, 1)
         XCTAssertEqual(got.first, .product("Dependency.framework"))
+        
+        let frameworkGot = try graph.linkableDependencies(path: project.path,
+                                                          name: dependency.name)
+        
+        XCTAssertEqual(frameworkGot.count, 2)
+        XCTAssertTrue(frameworkGot.contains(.product("libStaticDependency1.a")))
+        XCTAssertTrue(frameworkGot.contains(.product("libStaticDependency2.a")))
     }
```
Comment (kwridan) [inline Tests/TuistKitTests/Graph/GraphTests.swift]: Is it worth adding a test case to cover this scenario:

- App -> Framework1 (dynamic) -> Framework2 (static)

Verifying app only links (Framework1)

File: fixtures/ios_app_with_framework_linking_static_framework/App/Sources/AppDelegate.swift [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import Framework1
+import Framework2
```
Comment (kwridan) [inline fixtures/ios_app_with_framework_linking_static_framework/App/Sources/AppDelegate.swift]: Should the manifest be updated to declare there is a dependency on `Framework2` ?

File: Tests/TuistKitTests/Graph/GraphTests.swift (hunk 2 above)
Comment (ollieatkinson) [reply Tests/TuistKitTests/Graph/GraphTests.swift]: Good idea - will add this scenario

**Comment by ollieatkinson** [comment]
I added the following linter warning when linking any static product more than once:

```swift
"Target \(toTargetNode.target.name) has been linked against \(oldMember.targetNode.target.name) and \(targetNode.target.name), it is a static product so may introduce unwanted side effects."
```

**Comment by ollieatkinson** [comment]
>  Let's say an app links dynamically two frameworks, and the two of them link the same library statically. If that results in duplicated symbols, I'd add a linting step that checks that and alerts the user. 

Since the static products are not being linked into the same executable we don't get duplicated symbols. We do however end up in a weird scenario with multiple versions of the same type. This is valid in some scenarios for people's setup - I have however introduced a lint warning to let you know that you have linked the static product more than once.

File: fixtures/ios_app_with_framework_linking_static_framework/App/Sources/AppDelegate.swift (hunk 3 above)
Comment (ollieatkinson) [reply fixtures/ios_app_with_framework_linking_static_framework/App/Sources/AppDelegate.swift]: No, this import should be removed in this example.

**Pull request merged**